pkg/skuba/actions/addon/upgrade/plan.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2019 SUSE LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubeadm"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/upgrade/addon"
+	"github.com/SUSE/skuba/pkg/skuba"
+)
+
+func Plan() error {
+	fmt.Printf("%s\n", skuba.CurrentVersion().String())
+
+	client, err := kubernetes.GetAdminClientSet()
+	if err != nil {
+		return err
+	}
+	currentClusterVersion, err := kubeadm.GetCurrentClusterVersion(client)
+	if err != nil {
+		return err
+	}
+	currentVersion := currentClusterVersion.String()
+	latestVersion := kubernetes.LatestVersion().String()
+	allNodesVersioningInfo, err := kubernetes.AllNodesVersioningInfo()
+	if err != nil {
+		return err
+	}
+	allNodesMatchClusterVersion := kubernetes.AllNodesMatchClusterVersionWithVersioningInfo(allNodesVersioningInfo, currentClusterVersion)
```

Comment:
I think we could extract this logic because `skuba addon upgrade apply` will also need to perform this check, and this logic can be reused there. Can also be refactored on the `skuba addon upgrade apply` PR though.

Reply:
As it is just one function call, there's not so much to extract I'm afraid. Also we need `currentClusterVersion` in the plan output, so extracting this as well would mean we have to fetch it twice. I can only think of this but I think it provides not much added value:

```go
package addon

func SanityCheck(allNodesVersioningInfo kubernetes.NodeVersionInfoMap, currentClusterVersion *version.Version) bool {
	return kubernetes.AllNodesMatchClusterVersionWithVersioningInfo(allNodesVersioningInfo, currentClusterVersion)
}
```

Maybe you have a better idea there

Reply:
What I meant was to extract the whole piece:

```go
        currentClusterVersion, err := kubeadm.GetCurrentClusterVersion(client)
	if err != nil {
		return err
	}
	currentVersion := currentClusterVersion.String()
	allNodesVersioningInfo, err := kubernetes.AllNodesVersioningInfo()
	if err != nil {
		return err
	}
	allNodesMatchClusterVersion := 
        kubernetes.AllNodesMatchClusterVersionWithVersioningInfo(allNodesVersioningInfo, currentClusterVersion)
```

In a single function helper call, so `skuba addons upgrade apply` can use it as well.

Reply:
yes I understood, but we'd have to leave half of it

```go
         currentClusterVersion, err := kubeadm.GetCurrentClusterVersion(client)
	if err != nil {
		return err
	}
	currentVersion := currentClusterVersion.String()
```

because we need it in the `plan` output. so we'd have to fetch the cluster version twice, and extracting would not be beneficial at all. same for `apply` i guess

Reply:
I see, thanks for the explanation :)